File: src/tcp.py
"""Module responsible for handling TCP communication for a node.

Parses TCP requests provided by the browser and either initiates a request to a
node's neighbours or closes the TCP connection if an invalid request is made
"""
import socket
import re
from Node import Node
from FrameType import FrameType
from util import http_string
from Frame import Frame
from constants import MAX_INT
from udp import send_frame_to_neighbours
from timetable import check_timetable


def process_tcp(this_node: Node, transmission: bytes, sock: socket.socket) -> None:
    """Parses TCP requests sent by the browser.

    Uses regex to search for destination information and sends a request to a node's neighbours
    or closes the connection to the browser if the request is invalid.

    Arguments:
        this_node: instance of the node class representing a given node in the network.
        transmission: byte array containing the TCP packet
        sock: socket object representing the socket the TCP packet originated from
    Returns:
        None
    """
    transmission_str = transmission.decode("utf-8")
    destination = re.search(r"to=\w+", transmission_str)
    if not destination:
        response = http_string(400, "Bad Request", ["Ouch! Bad request"])
        sock.send(response.encode("utf-8"))
        this_node.input_sockets.remove(sock)
        sock.shutdown(socket.SHUT_RD)
        sock.close()
    else:

        destination = destination.group(0).split("=")[-1]
        request_frame = Frame(
            origin=this_node.name,
            dest=destination,
            src=[],
            seqno=this_node.seqno,
            time=-1,
            type=FrameType.REQUEST,
        )
        this_node.response_sockets[this_node.seqno] = sock
        this_node.seqno = (this_node.seqno + 1) % MAX_INT
        check_timetable(this_node)
        send_frame_to_neighbours(this_node, request_frame)

File: src/tcp.cpp
/*
Utility functions for handling TCP packets from the browser
*/
#include "tcp.h"

/*
Processes incoming TCP packets.

Parses packets using regex to check if they are valid requests, if they are
sends a requests to neighbours, elsewise it terminates the TCP connection with 
an appropriate message. 

Arguments:
  this_node: Node object representing station in the network
  message: the TCP packet transmitted by the browser
  socket: file descriptor of the socket that transmitted the request
Returns:
  void
*/
void handle_tcp(Node &this_node, string &message, int socket) {
  std::regex reg_str("to=\\w+");
  std::smatch match;
  std::regex_search(message, match, reg_str);
  string destination = match.str();
  if (match.size() == 0) {
    list<string> msgs({"Ouch! bad request"});
    string response = http_string(400, "Bad request", msgs);
    this_node.send_tcp(socket, response);
    this_node.remove_socket(socket);
    return;
  }
  size_t pos = destination.find("=");
  // Trim off prefix
  destination.erase(0, pos + 1);
  list<string> new_src;
  Frame request_frame =
      Frame(this_node.name, destination, new_src, this_node.seqno, -1, REQUEST);
  this_node.response_sockets[this_node.seqno] = socket;
  this_node.seqno = (this_node.seqno + 1) % MAX_INT;
  this_node.check_timetable();

  send_frame_to_neighbours(this_node, request_frame);
}